src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1212 @@
+                        .unwrap_or_else(|_| hint::unreachable_unchecked()),
+                );
+
+                // Copy the control bytes unchanged. We do this in a single pass
+                self.ctrl(0)
+                    .copy_to_nonoverlapping(new_table.ctrl(0), self.buckets() + Group::WIDTH);
+
+                {
+                    // The cloning of elements may panic, in which case we need
+                    // to make sure we drop only the elements that have been
+                    // cloned so far.
+                    let mut guard = guard((0, &mut new_table), |(index, new_table)| {
+                        if mem::needs_drop::<T>() {
+                            for i in 0..=*index {
+                                if is_full(*new_table.ctrl(i)) {
+                                    new_table.bucket(i).drop();
+                                }
+                            }
+                        }
+                        new_table.free_buckets();
+                    });
+
+                    for from in self.iter() {
+                        let index = self.bucket_index(&from);
+                        let to = guard.1.bucket(index);
+                        to.write(from.as_ref().clone());
+
+                        // Update the index in case we need to unwind.
+                        guard.0 = index;
+                    }
+
+                    // Successfully cloned all items, no need to clean up.
+                    mem::forget(guard);
+                }
+
+                // Return the newly created table.
+                new_table.items = self.items;
+                new_table.growth_left = self.growth_left;
+                ManuallyDrop::into_inner(new_table)
+            }
+        }
+    }
+}
+
+unsafe impl<#[may_dangle] T> Drop for RawTable<T> {
+    #[inline]
+    fn drop(&mut self) {
+        if self.bucket_mask != 0 {
+            unsafe {
+                if mem::needs_drop::<T>() {
+                    for item in self.iter() {
+                        item.drop();
+                    }
+                }
+                self.free_buckets();
+            }
+        }
+    }
+}
+
+impl<T> IntoIterator for RawTable<T> {
+    type Item = T;
+    type IntoIter = RawIntoIter<T>;
+
+    #[inline]
+    fn into_iter(self) -> RawIntoIter<T> {
+        unsafe {
+            let iter = self.iter();
+            let alloc = self.into_alloc();
+            RawIntoIter { iter, alloc }
+        }
+    }
+}
+
+/// Iterator over a a sub-range of a table. Unlike `RawIter` this iterator does
+/// not track an item count.
+pub struct RawIterRange<T> {
+    // Using *const here for covariance
+    data: *const T,
+    ctrl: *const u8,
+    current_group: BitMask,
+    end: *const u8,
+}
+
+impl<T> RawIterRange<T> {
+    /// Returns a `RawIterRange` covering a subset of a table.
+    ///
+    /// The start offset must be aligned to the group width.
+    #[inline]
+    unsafe fn new(
```

Comment:
It appears that the last Group of the range is fully processed even if it's partially truncated by range.end? If intended, this should be documented. Otherwise, it should be fixed.